File: api-clients/clojure/src/otreeba_open_cannabis_api/api/studies.clj
(ns otreeba-open-cannabis-api.api.studies
  (:require [otreeba-open-cannabis-api.core :refer [call-api check-required-params with-collection-format]])
  (:import (java.io File)))

(defn get-studies-with-http-info
  "Get a list of all current studies.
  Returns a paginated list of studies."
  ([] (get-studies-with-http-info nil))
  ([{:keys [page count sort ]}]
   (call-api "/studies" :get
             {:path-params   {}
              :header-params {}
              :query-params  {"page" page "count" count "sort" sort }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    ["api_key"]})))

(defn get-studies
  "Get a list of all current studies.
  Returns a paginated list of studies."
  ([] (get-studies nil))
  ([optional-params]
   (:data (get-studies-with-http-info optional-params))))

(defn get-studies-by-condition-with-http-info
  "Get a list of all current studies for a given condition.
  Returns a paginated list of studies."
  ([condition-slug ] (get-studies-by-condition-with-http-info condition-slug nil))
  ([condition-slug {:keys [page count sort ]}]
   (call-api "/studies/conditions/{conditionSlug}" :get
             {:path-params   {"conditionSlug" condition-slug }
              :header-params {}
              :query-params  {"page" page "count" count "sort" sort }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    ["api_key"]})))

(defn get-studies-by-condition
  "Get a list of all current studies for a given condition.
  Returns a paginated list of studies."
  ([condition-slug ] (get-studies-by-condition condition-slug nil))
  ([condition-slug optional-params]
   (:data (get-studies-by-condition-with-http-info condition-slug optional-params))))

(defn get-studies-conditions-with-http-info
  "Get a list of all current conditions for studies.
  Returns a list of all current conditions for studies."
  ([] (get-studies-conditions-with-http-info nil))
  ([{:keys [sort ]}]
   (call-api "/studies/conditions" :get
             {:path-params   {}
              :header-params {}
              :query-params  {"sort" sort }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    ["api_key"]})))

(defn get-studies-conditions
  "Get a list of all current conditions for studies.
  Returns a list of all current conditions for studies."
  ([] (get-studies-conditions nil))
  ([optional-params]
   (:data (get-studies-conditions-with-http-info optional-params))))

(defn get-study-by-identifier-with-http-info
  "Find study by DOI, PubMed ID, or slug.
  Returns a single study."
  [identifier-type identifier ]
  (call-api "/studies/{identifierType}/{identifier}" :get
            {:path-params   {"identifierType" identifier-type "identifier" identifier }
             :header-params {}
             :query-params  {}
             :form-params   {}
             :content-types ["application/json"]
             :accepts       ["application/json"]
             :auth-names    ["api_key"]}))

(defn get-study-by-identifier
  "Find study by DOI, PubMed ID, or slug.
  Returns a single study."
  [identifier-type identifier ]
  (:data (get-study-by-identifier-with-http-info identifier-type identifier)))

File: api-clients/elixir/lib/otreeba_open_cannabis_api/api/studies.ex
defmodule OtreebaOpenCannabisAPI.Api.Studies do
  @moduledoc """
  Documentation for OtreebaOpenCannabisAPI.Api.Studies.
  """

  use Tesla

  plug Tesla.Middleware.BaseUrl, "https://api.otreeba.com/v1"
  plug Tesla.Middleware.JSON

  @doc """
  Get a list of all current studies.

  Returns a paginated list of studies.
  """
  def get_studies(page, count, sort) do
    method = [method: :get]
    url = [url: "/studies"]
    query_params = [query: [{:"page", page}, {:"count", count}, {:"sort", sort}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  @doc """
  Get a list of all current studies for a given condition.

  Returns a paginated list of studies.
  """
  def get_studies_by_condition(condition_slug, page, count, sort) do
    method = [method: :get]
    url = [url: "/studies/conditions/#{condition_slug}"]
    query_params = [query: [{:"page", page}, {:"count", count}, {:"sort", sort}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  @doc """
  Get a list of all current conditions for studies.

  Returns a list of all current conditions for studies.
  """
  def get_studies_conditions(sort) do
    method = [method: :get]
    url = [url: "/studies/conditions"]
    query_params = [query: [{:"sort", sort}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  @doc """
  Find study by DOI, PubMed ID, or slug.

  Returns a single study.
  """
  def get_study_by_identifier(identifier_type, identifier) do
    method = [method: :get]
    url = [url: "/studies/{identifierType}/#{identifier}"]
    query_params = []
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end
end
